src/compiler/cse.rs

```diff
@@ -0,0 +1,452 @@
+
+        // Extract all the expressions from the step type
+        let mut exprs = Vec::<Expr<F, Queriable<F>, HashResult>>::new();
+
+        for constraint in &step_type_with_hash.constraints {
+            exprs.push(constraint.expr.clone());
+        }
+
+        // Find the optimal subexpression to replace
+        if let Some(common_expr) = find_optimal_subexpression(&exprs, &replaced_hashes, scorer) {
+            // Add the hash of the replaced expression to the set
+            replaced_hashes.insert(common_expr.meta().hash);
+            // Create a new signal for the common subexpression
+            let (common_se, decomp) = create_common_ses_signal(&common_expr, &mut signal_factory);
+
+            // Add the new signal to the step type and a constraint for it
+            decomp.auto_signals.iter().for_each(|(q, expr)| {
+                if let Queriable::Internal(signal) = q {
+                    step_type_with_hash.add_internal(*signal);
+                }
+                step_type_with_hash.auto_signals.insert(*q, expr.clone());
+                step_type_with_hash.add_constr(format!("{:?}", q), expr.clone());
+            });
+            decomp.constrs.iter().for_each(|expr| {
+                step_type_with_hash.add_constr(format!("{:?}", expr), expr.clone());
+            });
+
+            // Replace the common subexpression in all constraints
+            step_type_with_hash
+                .constraints
+                .iter_mut()
+                .for_each(|constraint| {
+                    constraint.expr = replace_expr(&constraint.expr, &common_se);
+                });
+        } else {
+            // No more common subexpressions found, exit the loop
+            break;
+        }
+        *step_type = step_type_with_hash.transform_meta(|_| ());
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(super) struct SubexprInfo {
+    count: usize,
+    degree: usize,
+}
+
+/// Information about a subexpression to help find the optimal subexpression to replace.
+impl SubexprInfo {
+    fn new(count: usize, degree: usize) -> Self {
+        Self { count, degree }
+    }
+
+    fn update(&mut self, degree: usize) {
+        self.count += 1;
+        self.degree = self.degree.max(degree);
+    }
+}
+
+/// Find the optimal subexpression to replace in a list of expressions.
+fn find_optimal_subexpression<F: Field + Hash, S: Scoring<F>>(
+    exprs: &[Expr<F, Queriable<F>, HashResult>],
+    replaced_hashes: &HashSet<u64>,
+    scorer: &S,
+) -> Option<Expr<F, Queriable<F>, HashResult>> {
+    let mut count_map = HashMap::<u64, SubexprInfo>::new();
+    let mut hash_to_expr = HashMap::<u64, Expr<F, Queriable<F>, HashResult>>::new();
+
+    // Extract all subexpressions and count them
+    for expr in exprs.iter() {
+        count_subexpressions(expr, &mut count_map, &mut hash_to_expr, replaced_hashes);
+    }
+
+    // Find the best common subexpression to replace based on the score
+    let best_subexpr = count_map
```

Comment:
If you iterate over the expressions instead of the count_map, you don't need hash_to_expr, because the expressions already contain the hash.

Reply:
But this `count_map` main purpose is to count the expressions and I think the counting should be outside of the expression itself because it should only live for each cse cycle. What I did was to store the needed information from the expression alongside with the count to be more straightforward